_posts/2021-01-04-Designing-REST-services.md

```diff
@@ -0,0 +1,331 @@
+</tbody>
+</table>
+
+<table>
+<tr><td><u>Sidebar: Singular versus Plural</u><br/>
+This book identifies resources using their singular noun in the URI. 
+Many service providers will use the plural noun instead. 
+There are many philosophical debates to be had whether to use singular or plural. 
+For example: Amazon and Google use plural in their API. 
+There is no clear reason or advantage for either to be preferred over the other, so in my opinion it is largely up to the designer(s) to choose as long as they remain consistent across the many components within the organization. 
+The only vaguely logical argument I found was that there are apparently more words in the English language, called mass nouns, that don’t have a plural form (such as “luggage”) than there are words that don’t have a singular form (such as “clothes”). 
+Bear in mind that I didn’t perform any due diligence of this argument, so the author of this reasoning might just be wrong about this.</td></tr>
+</table>
+
+When designing and implementing services we always need to consider whether calling these services should be safe and/or idempotent. 
+Safe services have no server-side side effects that the consumer of the service can be held accountable for. 
+These service calls will not trigger any effects that will change the state of resources. 
+Idempotency is the fact that a service call can be done multiple times without yielding a different result in any of its calls. 
+Each identical call will result in an identical response. 
+The GET of a resource is a call that is considered both safe and idempotent. 
+The service call will return the same result no matter how many times it is called, and it will not alter the state of the resource it is requesting.
+
+|HTTP Verb|Safe|Idempotent|
+|-------|-------|--------|
+|GET|x|x|
+|POST|x|x|
+|PUT||x|
+|DELETE||x|
+|PATCH|||
+
+Other verbs (all both safe and idempotent) that are commonly enabled on REST services are the following:
+* HEAD: Returns only the HTTP headers of the request in order to determine the context of the resource.
+* OPTIONS: Queries the endpoint for the possibilities it offers. This verb is typically also used by browsers as a preflight request to determine whether Cross-Origin Resource Sharing (CORS) is allowed
+* TRACE: Allows for a loop-back test with debug information. This method is not included in default authorization checking, and should be disabled in production, as it can be a security risk.
+
+The book concludes its elaboration on the second level of Richardson’s Maturity with an overview of some of the more popular technologies and frameworks and how they tackle the creation of a consumer for the services defined with REST. 
+These are rather straightforward for those developers that routinely use them, so we will not go deeper into them.
+
+## Level 3 - Hypermedia Controls
```

Comment:
Maybe something worth mentioning is that in order to successfully implement level 3 you really need to have a mature development team. It is pretty challenging which is why level 2 seems to be the most common for companies to pursuit. Also, if your clients don't make use of the features offered then it just creates a lot of overhead for the service developers.

Reply:
added